### DISCUSSION TITLE
Add a new deployment option for .Net Aspire

### DISCUSSION THREAD

**Discussion by hades200082** [coollabsio/coolify#2024] [Feature Requests]
.Net Aspire is a new development model Microsoft is pushing for .Net 8+ (which can run on Linux) that allows the project to manage the orchestration of services automatically. [Read more](https://learn.microsoft.com/en-us/dotnet/aspire/get-started/aspire-overview)

Another project called "Aspirate" is available that can take a .Net Aspire project and generate docker compose or kubernettes manifests from it. [Read more](https://github.com/prom3theu5/aspirational-manifests)

Together these could be used in a new build pack to build/publish .net Aspire application stacks along with spinning up containers for all required services. Aspire takes care of passing the connection strings and other details through to the various applications.

Supporting this would allow Coolify to get ahead of the game on containerised hosting and orchestration.

**Comment by coen22** [comment]
I tried it using aspirate, which can create a docker-compose file
https://github.com/coen22/AspireDockerComposeSample

The only thing that I get stuck on is that locally I run 'aspirate build' to push the images to Docker Desktop. This does not happen automatically in coolify as the generated docker compose is aimed at pushed images, instead of referencing Docker files.

I think Aspire is really cool, it works well on Azure, but none of this (docker stuff) is really supported and I ended up doing a lot of things manually which defeats the purpose.

This could be adapted though https://github.com/prom3theu5/aspirational-manifests to generate docker compose files suitable to coolify :)

**Reply by emimontesdeoca** [reply]
I did take a look at your source code as I tried basically the same, create a solution with .NET Aspire and somehow generate a docker-file.yml, then referencing it in a project in Coolify but had no look, I have problems with ports and stuff. Specially when using containers like Keycloak, or even normal APIs.

Did you manage to make it work?

I did try with your repo (https://github.com/coen22/AspireDockerComposeSample.git) and I get that error too, maybe I'm not doing it correctly?

I think supporting this in Coolify would be a gamechanger for people that skips cloud.

![image](https://github.com/user-attachments/assets/a200b4b8-c906-4a07-800b-543747e81313)

```
2024-Nov-19 11:36:56.592080
#36 exporting layers 0.0s done
2024-Nov-19 11:36:56.645018
#36 writing image sha256:8b1088a3d0285c5a984850e77479ca5279aee74fe01621ad80bf1468b01fd44c done
2024-Nov-19 11:36:56.645018
#36 naming to docker.io/library/xcw0ow8sskoskogggg8k8c0k-weatherapi done
2024-Nov-19 11:36:56.645018
#36 DONE 0.1s
2024-Nov-19 11:36:56.645018
2024-Nov-19 11:36:56.645018
#37 [weatherapi] resolving provenance for metadata file
2024-Nov-19 11:36:56.645018
#37 DONE 0.0s
2024-Nov-19 11:36:56.659063
Removing old containers.
2024-Nov-19 11:36:56.805512
Starting new application.
2024-Nov-19 11:36:57.605568
docker exec f4oo4000sc8808o88wggo8ck bash -c 'SOURCE_COMMIT=d9b3c1c9d5dae5f50bbb2dd44d8bcef61c067cc9 COOLIFY_BRANCH=main  docker compose --env-file /artifacts/f4oo4000sc8808o88wggo8ck/.env --project-name xcw0ow8sskoskogggg8k8c0k --project-directory /artifacts/f4oo4000sc8808o88wggo8ck -f /artifacts/f4oo4000sc8808o88wggo8ck/docker-compose.yaml up -d'
2024-Nov-19 11:36:57.605568
Container weatherapi-xcw0ow8sskoskogggg8k8c0k-113621252084  Creating
2024-Nov-19 11:36:57.605568
 Container react-xcw0ow8sskoskogggg8k8c0k-113621254093  Creating
2024-Nov-19 11:36:57.649445
Container react-xcw0ow8sskoskogggg8k8c0k-113621254093  Created
2024-Nov-19 11:36:57.649445
 Container weatherapi-xcw0ow8sskoskogggg8k8c0k-113621252084  Created
2024-Nov-19 11:36:57.655393
Container weatherapi-xcw0ow8sskoskogggg8k8c0k-113621252084  Starting
2024-Nov-19 11:36:57.655393
 Container react-xcw0ow8sskoskogggg8k8c0k-113621254093  Starting
2024-Nov-19 11:36:57.917643
Container react-xcw0ow8sskoskogggg8k8c0k-113621254093  Started
2024-Nov-19 11:36:57.922522
Error response from daemon: driver failed programming external connectivity on endpoint weatherapi-xcw0ow8sskoskogggg8k8c0k-113621252084 (58c0d5378b42a9e3ad1bc7b85a3a94315865b31ee9f5ab3411667f612a4dc331): Bind for 0.0.0.0:10001 failed: port is already allocated
2024-Nov-19 11:36:57.989723
Oops something is not okay, are you okay? 😢
2024-Nov-19 11:36:57.995506
Container weatherapi-xcw0ow8sskoskogggg8k8c0k-113621252084  Creating
2024-Nov-19 11:36:57.995506
 Container react-xcw0ow8sskoskogggg8k8c0k-113621254093  Creating
2024-Nov-19 11:36:57.995506
 Container react-xcw0ow8sskoskogggg8k8c0k-113621254093  Created
2024-Nov-19 11:36:57.995506
 Container weatherapi-xcw0ow8sskoskogggg8k8c0k-113621252084  Created
2024-Nov-19 11:36:57.995506
 Container weatherapi-xcw0ow8sskoskogggg8k8c0k-113621252084  Starting
2024-Nov-19 11:36:57.995506
 Container react-xcw0ow8sskoskogggg8k8c0k-113621254093  Starting
2024-Nov-19 11:36:57.995506
 Container react-xcw0ow8sskoskogggg8k8c0k-113621254093  Started
2024-Nov-19 11:36:57.995506
Error response from daemon: driver failed programming external connectivity on endpoint weatherapi-xcw0ow8sskoskogggg8k8c0k-113621252084 (58c0d5378b42a9e3ad1bc7b85a3a94315865b31ee9f5ab3411667f612a4dc331): Bind for 0.0.0.0:10001 failed: port is already allocated
```

**Reply by hades200082** [reply]
Aspirate did make some changes to support compose build from dokerfile. 

It might be possible now?

With aspire 9 they also ship an image for the aspire dashboard. 

When you deploy to azure with azd cli it also gives you that dashboard.

**Reply by coen22** [reply]
So, the thing is, it does deploy. Or it used to for me, at least. But the big problem that I still have is the automatic routing. How to make sure that the frontend knows where the backend is.

**Reply by coen22** [reply]
When using a blazor project with API and frontend. It puts those on the same port, so then it does work to deploy using a Dockerfile by pointing coolify to a Github repo. It automatically deploys everything. But what I did with that test is, I used an Sqlite database, which doesn't require a separate connection. Though I assume it could find a postgres database on its localhost when configured through aspire.

I've not deployed any aspire project to Azure itself, but that could explain something about how Microsoft configures those routes without having it be made explicit in the code (i.e. you could always define the api.example.com in the main aspire project).

**Reply by hades200082** [reply]
Deploying multiple apps like API + Blazor or even API + NextJS should be possible via aspire - you can do it with a docker compose so not sure why it would be a problem in aspire.

**Comment by hades200082** [comment]
Some other options for supporting native .NET deployments here might be:

- Add .NET 6-9 support to NixPacks
- Utilise .NET's built-in containerization publishing (you can publish a container without a dockerfile!)

**Comment by krdzex** [comment]
This would be huge for .NET devs. Big +1

**Comment by amirakrari** [comment]
Hi, would like to have some news regarding .NET Aspire deployment on coolify.
Does it work?

**Reply by hades200082** [reply]
> Hi, would like to have some news regarding .NET Aspire deployment on coolify.
> 
> Does it work?

You can use container deployment with Dockerfile, or docker compose with build already. 

There's no native hosting/deployment for .net in coolify. 

There's also no nixpack deployment for .net

**Reply by amirakrari** [reply]
Thanks for the answer!

**Comment by hades200082** [comment]
With the updates in Aspire 13.2 and the Aspire CLI, it might now be viable to have Coolify run the `aspire deploy` command directly (assuming the Aspire CLI is installed on the host machine and the Aspire apphost is configured for Docker Compose deployment)

1. Pull the repo
2. Execute `aspire deploy` in the root of the repo

This would trigger Aspire to build and publish Docker images for apps and produce a Docker Compose file, which it would then spin up itself.

There's also `aspire publish`, which just generates the Docker Compose file. 

If Coolify wanted to manage things like storage paths, etc. then it may need to parse the compose file and rewrite bits of it.

**Reply by hades200082** [reply]
There's also https://github.com/[quentin-r37/aspire-to-coolify](https://github.com/quentin-r37/aspire-to-coolify) which uses the Coolify API to handle deployments in a more AZD style way.
